kernel/src/table_features/column_mapping.rs

```diff
@@ -56,3 +59,129 @@ impl AsRef<str> for ColumnMappingMode {
         }
     }
 }
+
+/// Extract the schema and column mapping mode from the metadata. When column mapping mode is
+/// enabled, verify that each field in the schema is annotated with a physical name and field_id;
+/// when not enabled, verify that no fields are annotated.
+pub(crate) fn get_validated_column_mapping_schema(
+    metadata: &Metadata,
+    protocol: &Protocol,
+) -> DeltaResult<(Schema, ColumnMappingMode)> {
+    let column_mapping_mode = match metadata.configuration.get(COLUMN_MAPPING_MODE_KEY) {
+        Some(mode) if protocol.min_reader_version() >= 2 => mode.as_str().try_into()?,
+        _ => ColumnMappingMode::None,
+    };
+    let schema = metadata.schema()?;
+    validate_column_mapping_schema(&schema, column_mapping_mode)?;
+    Ok((schema, column_mapping_mode))
+}
+
+pub(crate) fn validate_column_mapping_schema(
+    schema: &Schema,
+    mode: ColumnMappingMode,
+) -> DeltaResult<()> {
+    if mode == ColumnMappingMode::Id {
+        // TODO: Support column mapping ID mode
+        return Err(Error::unsupported("Column mapping ID mode not supported"));
+    }
+
+    let mut validator = ValidateColumnMappings {
+        mode,
+        path: vec![],
+        err: None,
+    };
+    let _ = validator.transform_struct(schema);
+    if let Some(err) = validator.err {
+        Err(err)
+    } else {
+        Ok(())
+    }
+}
+
+struct ValidateColumnMappings {
+    mode: ColumnMappingMode,
+    path: Vec<String>,
+    err: Option<Error>,
+}
+impl ValidateColumnMappings {
+    fn transform_inner_type<'a>(
+        &mut self,
+        data_type: &'a DataType,
+        name: &str,
+    ) -> Option<Cow<'a, DataType>> {
+        if self.err.is_none() {
+            self.path.push(name.to_string());
+            let _ = self.transform(data_type);
+            self.path.pop();
+        }
+        None
+    }
+    fn check_annotations(&mut self, field: &StructField) {
+        let annotation = "delta.columnMapping.physicalName";
+        match (self.mode, field.metadata.get(annotation)) {
+            // Both Id and Name modes require a physical name; None mode requires no physical name.
+            (ColumnMappingMode::Name | ColumnMappingMode::Id, Some(MetadataValue::String(_))) => {}
+            (ColumnMappingMode::None, None) => {}
+            (ColumnMappingMode::None, Some(_)) => {
+                self.err = Some(Error::invalid_column_mapping_mode(format!(
+                    "Column mapping is not enabled but field '{}' is annotated with {}",
+                    ColumnName::new(&self.path),
+                    annotation
+                )));
+            }
+            (ColumnMappingMode::Name | ColumnMappingMode::Id, _) => {
```

Comment:
with the above change for a better error this becomes:
```suggestion
            (ColumnMappingMode::Name | ColumnMappingMode::Id, None) => {
```